### PR TITLE
Login API

### PR DESCRIPTION
**Pull request by prakhar1912**
I have added code for the login API. There is an issue with the tests. I would suggest taking a quick look at the code before reading further.

To test the login API, I have to create a mock user first.
There are 2 approaches to this:
1) Create one using the User model
2) Make a request to `/api/signup` using `supertest`

Approach 2 is running the tests fine, while Approach 1 is not creating the user.
I think Approach 1 has some issue with Mockgoose.

Fixes I have tried:
1) Create user in `beforeAll` and `beforeEach`
2) Use `User.create` instead of `new User()`
3) Make code sync by using `done` as a callback

### CODE REVIEW

File: server/__tests__/loginAPI.test.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    request(testServer)
+      .post('/api/login')
+      .send({
+        email,
+        password,
+      })
+      .expect(200)
+      .end((error, response) => {
+        const createdUser = response.body;
+        expect(createdUser).to.have.keys('id', 'username');
```
Comment (vkbinfo) [inline server/__tests__/loginAPI.test.js]: Shouldn't we check username also? to be sure that it's returning the username that we sent in the request.

File: server/__tests__/loginAPI.test.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    const { password } = mockUser;
+    request(testServer)
+      .post('/api/login')
+      .send({
+        email: 'thismailcannotexist@nomailaddress.com',
+        password,
+      })
+      .expect(400)
+      .end((error, response) => {
+        expect(response.body).to.have.key('errors');
```
Comment (vkbinfo) [inline server/__tests__/loginAPI.test.js]: is errors an array? if not, we should use error instead of errors.

Comment (niinpatel) [reply server/__tests__/loginAPI.test.js]: @vkbinfo , yes, errors is an array.

File: server/__tests__/loginAPI.test.js (hunk 1 above)
Comment (prakhar1912) [reply server/__tests__/loginAPI.test.js]: We are logging in using email right now. Logging in with username can be added later.

**Comment by prakhar1912** [comment]
> I have added code for the login API. There is an issue with the tests. I would suggest taking a quick look at the code before reading further.
> 
> To test the login API, I have to create a mock user first.
> There are 2 approaches to this:
> 
> 1. Create one using the User model
> 2. Make a request to `/api/signup` using `supertest`
> 
> Approach 2 is running the tests fine, while Approach 1 is not creating the user.
> I think Approach 1 has some issue with Mockgoose.
> 
> Fixes I have tried:
> 
> 1. Create user in `beforeAll` and `beforeEach`
> 2. Use `User.create` instead of `new User()`
> 3. Make code sync by using `done` as a callback

This has been resolved.

**Comment by niinpatel** [comment]
LGTM

File: server/src/controllers/userController.js [hunk 3]
```diff
@@ -22,6 +23,34 @@ const createNewUser = async (req, res) => {
   }
 };
 
+const loginUser = async (req, res) => {
+  const { email, password } = req.body;
+
+  try {
+    const user = await User.findOne({ email });
```
Comment (nimish-gupta) [inline server/src/controllers/userController.js]:
```suggestion
    const user = await User.findOne({ email });
    
    if (!user) {
        return res.status(400).send({
           errors: 'No user record found',
        });
    }

    const passwordMatch = await bcrypt.compare(password, user.password);

     if (!passwordMatch) {
        return res.status(401).send({
          errors: 'Invalid password',
         });
      }
      
      const token = user.generateAuthToken();
      res.set('X-Auth-Token', token);
      return res.send(user);
      
```

Comment (prakhar1912) [reply server/src/controllers/userController.js]: @nimish-gupta Good catch. I will make this change.

Comment (prakhar1912) [reply server/src/controllers/userController.js]: @nimish-gupta The changes have been made.

**Pull request merged**